test/helm-test/main.go

```diff
@@ -0,0 +1,211 @@
+	terminateTimeout = time.Duration(5 * time.Minute) // terminate 5 minutes after SIGINT is sent.
+
+	interrupt = time.NewTimer(interruptTimeout) // interrupt testing at this time.
+	terminate = time.NewTimer(time.Duration(0)) // terminate testing at this time.
+
+	suite TestSuite
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+var letterRunes = []rune("abcdefghijklmnopqrstuvwxyz")
+
+// return cmd.Output(), potentially timing out in the process.
+func output(cmd *exec.Cmd) ([]byte, error) {
+	interrupt.Reset(interruptTimeout)
+	stepName := strings.Join(cmd.Args, " ")
+	cmd.Stderr = os.Stderr
+
+	log.Printf("Running: %v", stepName)
+	defer func(start time.Time) {
+		log.Printf("Step '%s' finished in %s", stepName, time.Since(start))
+	}(time.Now())
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	type result struct {
+		bytes []byte
+		err   error
+	}
+	finished := make(chan result)
+	go func() {
+		b, err := cmd.Output()
+		finished <- result{b, err}
+	}()
+	for {
+		select {
+		case <-terminate.C:
+			terminate.Reset(time.Duration(0)) // Kill subsequent processes immediately.
+			syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
+			cmd.Process.Kill()
+			return nil, fmt.Errorf("Terminate testing after 15m after %s timeout during %s", interruptTimeout, stepName)
+		case <-interrupt.C:
+			log.Printf("Interrupt testing after %s timeout. Will terminate in another %s", interruptTimeout, terminateTimeout)
+			terminate.Reset(terminateTimeout)
+			if err := syscall.Kill(-cmd.Process.Pid, syscall.SIGINT); err != nil {
+				log.Printf("Failed to interrupt %v. Will terminate immediately: %v", stepName, err)
+				syscall.Kill(-cmd.Process.Pid, syscall.SIGTERM)
+				cmd.Process.Kill()
+			}
+		case fin := <-finished:
+			return fin.bytes, fin.err
+		}
+	}
+}
+
+func randStringRunes(n int) string {
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letterRunes[rand.Intn(len(letterRunes))]
+	}
+	return string(b)
+}
+
+func main() {
+	matches, err := filepath.Glob("stable/*")
+	defer writeXML("/workspace/_artifacts", time.Now())
+	if !terminate.Stop() {
+		<-terminate.C // Drain the value if necessary.
+	}
+
+	if !interrupt.Stop() {
+		<-interrupt.C // Drain value
+	}
+
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	for _, dir := range matches {
+		ns := randStringRunes(10)
+		rel := randStringRunes(3)
+
+		xmlWrap(fmt.Sprintf("lint %s", dir), func() error {
+			_, execErr := output(exec.Command("linux-amd64/helm", "lint", dir))
+			return execErr
+		})
+
+		xmlWrap(fmt.Sprintf("install %s", dir), func() error {
+			o, execErr := output(exec.Command("linux-amd64/helm", "install", dir, "--namespace", ns, "--name", rel, "--wait"))
+			if execErr != nil {
+				return fmt.Errorf("%s Command output: %s", execErr, string(o[:]))
+			}
+			return nil
+		})
+
+		xmlWrap(fmt.Sprintf("test %s", dir), func() error {
+			o, execErr := output(exec.Command("linux-amd64/helm", "test", rel))
```

Comment:
This command exits non-zero if a chart doesn't have any tests, is that handled for?